File: daily/53-maximum-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
  nums_length = nums.length
  max_sofar, max_ending = nums[0], nums[0]
  i = 1
  while i < nums_length
    max_ending = (max_ending + nums[i] > nums[i] ? max_ending + nums[i] : nums[i])
    max_sofar = (max_ending > max_sofar ? max_ending : max_sofar)
    i += 1
  end

  max_sofar
end

File: daily/53-maximum-subarray.go
func maxSubArray(nums []int) int {
    numsLen := len(nums)
    if numsLen == 0 {
        return 0
    }
    if numsLen == 1 {
        return nums[0]
    }
    
    var maxEnd, maxSoFar = nums[0], nums[0]
    
    for i:=1; i<numsLen; i++ {
        if maxEnd < 0 && maxEnd < nums[i] {
            maxEnd = nums[i]
        } else {
            maxEnd += nums[i]
        }

        if maxEnd > maxSoFar {
            maxSoFar = maxEnd
        }
    }
    
    return maxSoFar
}